File: frontend/src/hooks/rtcmessage.ts
export type RTCOffer = {
  type: 'rtc_offer';
  id: string;
  offer: RTCSessionDescriptionInit;
};
export type RTCAnswer = {
  type: 'rtc_answer';
  id: string;
  answer: RTCSessionDescriptionInit;
};
export type RTCIceCandidate = {
  type: 'rtc_ice_candidate';
  id: string;
  candidate: RTCIceCandidateInit;
};
export type RTCMessage = RTCOffer | RTCAnswer | RTCIceCandidate;

File: src/main/java/se/su/dsv/proctoring/web/proctor/RTCMessage.java
package se.su.dsv.proctoring.web.proctor;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

import java.util.UUID;

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = RTCMessage.Answer.class, name = "rtc_answer"),
        @JsonSubTypes.Type(value = RTCMessage.ICECandidate.class, name = "rtc_ice_candidate"),
        @JsonSubTypes.Type(value = RTCMessage.Offer.class, name = "rtc_offer"),
})
public interface RTCMessage {
    /**
     * The unique identifier for the connection this message is bound for.
     * @return the connection id
     */
    UUID connectionId();

    record Offer(
            @JsonProperty("id") UUID connectionId,
            @JsonProperty("offer") RTCSessionDescription offer)
            implements RTCMessage, CandidateMessage.Inbound, InboundMessage
    {}

    record Answer(
            @JsonProperty("id") UUID connectionId,
            @JsonProperty("answer") RTCSessionDescription answer)
            implements RTCMessage, CandidateMessage.Inbound, InboundMessage
    {}

    record ICECandidate(
            @JsonProperty("id") UUID connectionId,
            @JsonProperty("candidate") RTCIceCandidate candidate)
            implements RTCMessage, CandidateMessage.Inbound, InboundMessage
    {}
}
